### ISSUE TITLE
Improve the usability of keywords for tagging

### ISSUE THREAD

**Issue by claell** [JabRef/jabref#8739]
**Is your suggestion for improvement related to a problem? Please describe.**
Currently, there seem to be two options for tagging: groups and keywords. Currently, groups seem to be the most usable option. However, I wonder, whether keywords are actually the better way to tag documents. However, their current usability is very bad (for example, showing all documents with a keyword requires some search skills).

**Describe the solution you'd like**
Have keywords act like tags that can easily be used, possibly extending groups.

**Additional context**
-

**Comment by ThiloteE** [comment]
I assume, you want to have a better UI for keywords? Could you provide some examples how it should look like?
Currently JabRef provides keyword fields in some entry editor tab.

**Comment by Siedlerchr** [comment]
Otherwise you can also create automatic groups based on keywords, if that helps you

**Comment by claell** [comment]
Yes, I am talking about a better UI for keywords, I know about both, the option to edit them and the option to create groups based on keywords.

What I was thinking of is something like labels here on GitHub (how you add them, how they display on issues and how you can click them or search for them). If needed, I can add more detailed explanation with screenshots.

Possibly, there are even better ways (for the context of reference management) to provide a keyword UI, but that is what I can think of right now.

**Comment by claell** [comment]
When searching for keywords, I was more thinking about GitLab, than GitHub, actually:

![grafik](https://user-images.githubusercontent.com/6171135/166505421-1ed93d61-ac22-492a-9d83-6ba10c675864.png)

**Comment by Beingmani** [comment]
Hello, 

Can someone help me to replicate this screen in jabref, I have a solution that I can propose to improve the usability of the keywords. I have the design ready. I just need to check how it will look in jabref.

**Comment by ThiloteE** [comment]
@Beingmani, Claell did not share this screen from within JabRef, but this is how GitLab implemented keywords. It was just an example of how it could be.

**Comment by ThiloteE** [comment]
This is how JabRef currently implements keywords:

![grafik](https://user-images.githubusercontent.com/73715071/167808578-aff32bbf-0356-4ba3-816f-86f1b31a0611.png)

You then can use the search bar and search and search for specific keywords (e.g. "test") by writing `keywords=test`.

**Comment by ThiloteE** [comment]
Very much related: #7423

@mchellelm, how are you faring so far with your implementation?

**Comment by ThiloteE** [comment]
What I would like to see is keywords showing directly in the maintable (or in a separate specific area in the UI that is easily accessible and then I want to be able to click on it, which will lead to a place that shows me all my entries with this specific keyword. Just like I can do it here on GitHub with JabRef issues.

**Comment by ThiloteE** [comment]
Thank you mlep. This was an important note. I think there should be a separation between `keywords` and `labels`. Labels should make it easier for the user to "manage" the entries. The BibLaTeX field "Keywords" does too, but there is an important distinction: "keywords" may potentially be parsed with TeX engines and printed in the bibliography. "Labels" would not, as bib(la)tex has not defined that field, right?

Therefore, I propose to introduce the "Labels" field to JabRef and it to be JabRef exclusive (comparable to the groups field)

**Comment by mlep** [comment]
Being a special field in biblatex, the content of `keywords`  is "usually not printed" (according to the doc). A user could decide otherwise by altering a citation style (like with any other field), but I guess we should not take it into account.

I believe having `labels`, `groups` and `keywords` will be mostly redundant. Before defining the field `labels`, we should define the specific use of each these fields.
We could display keywords (or groups) as a label (like in github). In fact there is already something like this in JabRef: if a group as a color, the column "groups" of the entry editor shows a bar with this color.

**Comment by Beingmani** [comment]
Labels are mainly user-specific, which are more or less similar to Groups in general. It's like Grouping entries under a similar label. isn't?

And keywords are more specific to entries, Searching for a particular keywords popups the entries related to it. I have a rough sketch of how we can use keywords, let me share it here.

**Comment by claell** [comment]
@mlep Thanks for the insights. I had a look at page 333 of the biblatex docs, and it looks as if they are talking about different groups (groups of commands?) than what we do here. So not sure whether that is really related.

I agree that there is overlapping between `keywords` (possibly `labels`) and `groups`. Citavi also has both, keywords and groups AFAIK (although their groups are less versatile than the ones in JabRef).

In general, `groups` tend to "contain" entries, whereas `keywords` or `labels` tend to be assigned to entries.

**Comment by mlep** [comment]
> Labels are mainly user-specific, which are more or less similar to Groups in general. It's like Grouping entries under a similar label. isn't?

To me: yes it is, meaning we do not need to introduce the field `labels` since we have the field `groups`.
 
> And keywords are more specific to entries, Searching for a particular keywords popups the entries related to it. I have a rough sketch of how we can use keywords, let me share it here.

Another feature related to keywords: currently they can be displayed in the entry editor (as a column).

**Comment by mlep** [comment]
> @mlep Thanks for the insights. I had a look at page 333 of the biblatex docs, and it looks as if they are talking about different groups (groups of commands?) than what we do here. So not sure whether that is really related.

Yes, the name is the same, but the concept is different. So, I do not think there could a confusion between groups in JabRef and groups in biblatex.

> I agree that there is overlapping between `keywords` (possibly `labels`) and `groups`. Citavi also has both, keywords and groups AFAIK (although their groups are less versatile than the ones in JabRef).
> 
> In general, `groups` tend to "contain" entries, whereas `keywords` or `labels` tend to be assigned to entries.

I agree with your view: the fields `groups` and `keywords` have separated usage. What I do not see at the moment: what usage of a field `labels` that is not already ensured by the fields `groups` and `keywords`?

**Comment by claell** [comment]
@mlep I think, @ThiloteE suggested `labels` as an additional, JabRef exclusive field to not mess with the `keywords` field (to not cause unwanted side effects). However, if I read your comments above correctly, there most likely won't be side effects, which is why one can stay with `keywords` for this purpose.

**Comment by claell** [comment]
One possibly unwanted side effect that I can think of would be if keywords are also used in DOI or other places (I know that Citavi automatically sets keywords (from a source I currently don't know) when adding an entry). In that case, "official" keywords might conflict with "user" keywords. But not sure whether this really is (or can be) the case.

**Comment by ThiloteE** [comment]
> Being a special field in biblatex, the content of keywords is "usually not printed" (according to the doc).

Very good.

The only **sideeffect** I can think of now:

Adding keywords and then exporting this bibliographic data to other people will lead to **them having my "personal" keywords** I use to manage stuff in their library. By the way, this is also a problem for groups and attached files, but for those, it is easier, because I can just copy the entries to a new library and then to remove all group and file fields from the library. I can easily clean my library. For keywords this approach would not be possible. If I wanted to only export the keywords that come with importing from ISBN / DOI or some other identifier, I would need to go through all the entries manually.

**Comment by Siedlerchr** [comment]
And to fully mess up the distinction, you can also create automatic groups by keywords

**Comment by claell** [comment]
Agreed.

Based on that, I think that the UX of the possible solutions should be discussed (for example possible conflicts for `keyword` usage, etc.) to select a good path forward.

**Comment by ThiloteE** [comment]
@mlep I like the current maintable view a lot. I am not so much a fan of the proposed redesign (because of https://github.com/JabRef/jabref/issues/6857#issuecomment-762470398 and because one probably would have a harder time sorting and finding entries. The excel like view really helps a lot for sorting and ordering), so naturally I personally would prefer this to be added to the *current* maintable. Others might be of different opinion and that is fine. How about both? :)

@claell To be honest, I think this issue here does not need a lot more discussion, I think we mentioned the most advantages and disadvantages. We just need somebody who implements it. Further discussion can be done in the actual pull request. The longer the discussion goes, the more bothersome for someone who actually wants to work on this because they have to read all this stuff.

**Comment by mlep** [comment]
> @mlep I like the current maintable view a lot. I am not so much a fan of the proposed redesign (because of [#6857 (comment)](https://github.com/JabRef/jabref/issues/6857#issuecomment-762470398) and because one probably would have a harder time sorting and finding entries. The excel like view really helps a lot for sorting and ordering), so naturally I personally would prefer this to be added to the _current_ maintable. Others might be of different opinion and that is fine. How about both? :)

I am quite happy with the current maintable too. My concern is about "1. Show `groups`, `keywords` (and potentially `labels`) somewhere in the maintable.": will there be enough room? For example, displaying the keywords in the maintable (with the current design) do not allow to see them when except the column is pretty wide (in my databases, the list of official keywords is often longer than the title of the paper). Could the label be displayed on multiple lines for each entry?

**Comment by claell** [comment]
@ThiloteE regarding discussion: I think there are open points needing decisions, but yes, they can be discussed during implementation phase.

Personally, I don't like the current table view that much, so thanks for the reference of the redesigned maintable issue @mlep :)
Some ideas seem to make most sense with that, but maybe they can also be transferred to the current table design.

**Comment by mchellelm** [comment]
> @mchellelm, how are you faring so far with your implementation?

Hi, we've struggled a little bit in terms of understanding the code and just navigating it. One of my team members has a few questions he'll add to #7423 shortly, so any help with that would be appreciated.

**Comment by sjb-ch1mp** [comment]
G'day! I can see that there's been some discussion in this thread, but that the issue is currently unassigned. If possible, I would like to have a stab at this because it sounds like an interesting problem. I have had a preliminary look at the code base and I have **some** confidence that I might be concoct a solution. I will be happy to submit a draft pull request when I've made some progress so that you can review the direction that I'm headed in, if you'd prefer. I will be sure to take #7423 into account when I'm devising a way forward.

**Comment by ThiloteE** [comment]
Sure! You are welcome! :-)

As a general advise for JabRef newcomers: Please check out https://github.com/JabRef/jabref/blob/main/CONTRIBUTING.md for a start. Also, https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace is worth having a look at. Feel free to ask if you have any questions here on GitHub or also at JabRef's [Gitter](https://gitter.im/JabRef/jabref) chat.

**Comment by sjb-ch1mp** [comment]
Apologies @ThiloteE, I think I bit off more than I can chew with this one. I'm going to unassign myself so I don't waste anymore time. I will look for a more 'newcomer-friendly' issue, because this is far more complex a problem than I first anticipated.

**Comment by ThiloteE** [comment]
Fair enough.

For anybody else, who might want to have a shot at this: It's not like the issue has to be complete from the start. We have split the task up into multiple smaller issues (and if necessary could be split up into even smaller ones by the people that know more about the code) and it is good enough, if only one or a few of these smaller issues are worked on. Step by step :-)

**Comment by calixtus** [comment]
Lowered priority of this issue, as this has some larger preconditions, that have to be met first.
- https://github.com/JabRef/jabref/issues/6857
- https://github.com/JabRef/jabref/pull/8963
- https://github.com/koppor/jabref/issues/341 or https://github.com/JabRef/jabref/issues/7423

**Comment by ryan-carpenter** [comment]
> While BibTeX does not define the field `keywords`, biblatex does (as a special field). ... It specifies the comma as a separating character (Note: JabRef allows for other separating characters to be defined in the preferences).

Note: Biblatex also allows alternative separators, which is helpful when the keywords contain meaningful commas, as in chemical numbers and MEDLINE subject headings.

> **Separated value fields** A separated list of literal values. Examples are the keywords and options fields. The separator can be configured to be any Perl regular expression via the xsvsep option which defaults to the usual BibTeX comma surrounded by optional whitespace.

(from https://mirrors.ibiblio.org/CTAN/macros/latex/contrib/biblatex/doc/biblatex.pdf)

**Comment by shivam-gujar** [comment]
Hi, I’m Shivam. I’m learning Java and interested in improving keyword tagging usability. Could I work on this issue?

**Comment by koppor** [comment]
@CodebyShiv No - see that message: https://github.com/JabRef/jabref/issues/8739#issuecomment-1538920960

Look around for other issues - maybe you will also find something at https://github.com/JabRef/jabref-koppor/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen
